dali/operators/reader/numpy_reader_gpu_op.cc

```diff
@@ -85,15 +88,30 @@ void NumpyReaderGPU::Prefetch() {
   curr_tensor_list.Resize(tmp_shapes, ref_type);
 
   // read the data
+  int first_padded = curr_tensor_list.num_samples();
   for (int data_idx = 0; data_idx < curr_tensor_list.num_samples(); ++data_idx) {
     curr_tensor_list.SetMeta(data_idx, curr_batch[data_idx]->meta);
     SampleView<GPUBackend> sample(curr_tensor_list.raw_mutable_tensor(data_idx),
                                   curr_tensor_list.tensor_shape(data_idx),
                                   curr_tensor_list.type());
-    ScheduleChunkedRead(sample, *curr_batch[data_idx]);
+    // when padding the last sample is duplicated so no need to redo the same work
+    if (data_idx > 0 && curr_batch[data_idx -1 ] == curr_batch[data_idx]) {
+      first_padded = data_idx;
+    } else {
+      ScheduleChunkedRead(sample, *curr_batch[data_idx]);
+    }
   }
   thread_pool_.RunAll();
   staging_.commit();
+
+  // copy padded samples
+  for (int data_idx = first_padded; data_idx < curr_tensor_list.num_samples(); ++data_idx) {
+    kernels::copy<StorageGPU, StorageGPU>(curr_tensor_list.raw_mutable_tensor(data_idx),
+                                          curr_tensor_list.raw_tensor(data_idx - 1),
+                                          curr_tensor_list.tensor_offset(data_idx) -
+                                            curr_tensor_list.tensor_offset(data_idx - 1),
+                                          staging_stream_);
+  }
```

Comment:
I think this should rather be a scatter-gather operation (with overlapping source).

Reply:
Will do.